proposals/normalize-addressing.md

```diff
@@ -0,0 +1,175 @@
+and any other information relevant for prioritizing this project against others.
+It does not need to describe the work in much detail. Most technical design and planning would take place after a proposal is adopted.
+Good project scope aims for ~3-5 engineers for 1-3 months (though feel free to suggest larger-scoped projects anyway). 
+Projects do not include regular day-to-day maintenance and improvement work, e.g. on testing, tooling, validation, code clarity, refactors for future capability, etc.
+-->
+<!--
+For ease of discussion in PRs, consider breaking lines after every sentence or long phrase.
+-->
+
+## Purpose &amp; impact 
+#### Background &amp; intent
+_Describe the desired state of the world after this project? Why does that matter?_
+<!--
+Outline the status quo, including any relevant context on the problem you’re seeing that this project should solve. Wherever possible, include pains or problems that you’ve seen users experience to help motivate why solving this problem works towards top-line objectives. 
+-->
+
+Raising popularity of NFTs triggered debates how NFTs stored on IPFS should be addressed. On one hand links like `ipfs://${cid}/` are more future proof and decentralized, but on the other hand chances are high that such links in the wild will not be clickable and can harm adoption.
+
+Things are complicated further by a fact that same content can be addresed in several ways even on the same gateway. E.g. here is all the various ways [QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR](https://explore.ipld.io/#/explore/QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR) could be encountered by the users
+
+
+1. `ipfs://QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR`
+
+   > Which should be invalid and really should be following instead
+   >
+   > `ipfs://bafybeigdyrzt5sfp7udm7hu76uh7y26nf3efuylqabf3oclgtqy55fbzdi`
+
+2. https://bafybeigdyrzt5sfp7udm7hu76uh7y26nf3efuylqabf3oclgtqy55fbzdi.ipfs.dweb.link/
+
+3. https://dweb.link/ipfs/QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR
+
+4. https://dweb.link/ipfs/bafybeigdyrzt5sfp7udm7hu76uh7y26nf3efuylqabf3oclgtqy55fbzdi
+
+5. https://dweb.link/ipfs/?uri=ipfs://QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR
+
+
+All but `ipfs://` URL will get normalized into 2nd from, that is good but it also different enough from native `ipfs://` URL to make it confusing / inconvenient to work with.
+
+This proposal suggests to address those pain points through following:
+
+1. Making `ipfs://${CIDv1}` canonical representation.
+   > 
+   > We can drive adoption by adopting these representation across our libraries [like propsed here](https://github.com/multiformats/js-multiformats/issues/71)
+   > 
+
+2. Ensure that users see `ipfs://${CIDv1}` regardless of gateway they're on.
+   > Above examples from 2nd to 5th could redirect to:
+   > `ipfs://${originHash(CID)}.dweb.link/ipfs://${CIDv1}/`
```

Comment:
No cigar. Subdomains are how we want people to load websites because content root is aligned with URL root.
Adding this artificial path prefix breaks that alignment and breaks relative paths on every website :(

Reply:
> No cigar. Subdomains are how we want people to load websites because content root is aligned with URL root.

My understanding is we want to make sure that content is origin isolated based on CID. We can accomplish that without teaching people how to translate CIDs into origins etc... It is a lot easier to teach programs like gateways to do that, for humans we could just keep `ipfs://${CIDv1}/` regardless of (sub)domain. That gives us:

1. Origin base content separation we seek.
2. We retain same path in form of `ipfs://${CIDv1}` across all gateways. Reinforcing  that `${bridge}` in `${bridge}ipfs://${CIDv1}` is just a bridge to `ipfs://${CIDv1}`.

> Adding this artificial path prefix breaks that alignment and breaks relative paths on every website :(

You mean sites that use `/images/cat.png` would be broken ? That is a fare point, although I think something that we can overcome by redirecting to `/ipfs://{CIDv1}/images/cat.png` based on referrer.

Also would not those sites be as broken when loaded from `https://ipfs.io/ipfs/${CIDv1}` now ?

Reply:
Correct. We spent years fixing bugs in various static website generators to support truly relative paths, and it remained a big adoption blocker because there was always a new bug for something new filled.

Only when subdomain gateways got introduced, and the CID root became the root of URL path, all websites and tools started working out-of-the-box with IPFS. 

:point_right:  Redirects are transparent in browser context, so your fix will work for a subset of problems like loading images or other assets, but it won't fix websites that execute JS code based on  `window.location.pathname`. 

Due to this I don't believe this is feasible.

Reply:
> 👉 Redirects are transparent in browser context, so your fix will work for a subset of problems like loading images or other assets, but it won't fix websites that execute JS code based on `window.location.pathname` 💔

That historical context is really helpful! Do you by a chance have a pointers to above JS problems, so it could be quantified and pros and cons could be evaluated.

I realize this is an unfortunate tradeoff. Then I also think that turning `ipfs://{cid}/path/file` into `https:///${cid}.gateway.tld/path/file` is a cause of why later form is more prominent. I also believe that later form makes it really difficult for an average person to realize that gateway could be swapped.

Maybe there is a compromise, that make better tradeoffs ?